File: 2018/solutions/solution_05.py
import re

with open("./2018/input/input_05.txt") as f:
    input = f.read().strip()

alpha = "abcdefghijklmnopqrstuvwxyz"
pattern = []

for i in alpha:
    pattern.append(i + i.upper())
    pattern.append(i.upper() + i)

pattern = re.compile("|".join(pattern))

def react(poly):
    out = re.sub(pattern, "", poly)
    while out != poly:
        poly = out
        out = re.sub(pattern, "", poly)
    return out

# Part 1
def part1():
    return len(react(input))

# Part 2
def part2():
    reduced = []
    for i in alpha:
        letter = re.compile(f"{i}|{i.upper()}");
        reducePoly = re.sub(letter, "", input)
        reduced.append(len(react(reducePoly)))
    return min(reduced)

File: 2017/solutions/solution_05.js
const fs = require("fs");
const input = fs.readFileSync(__dirname + "/../input/input_05.txt", "utf8")
  .trim().split("\n").map(j => parseInt(j, 10));

// Part 1
function part1() {
  let hops = input.slice();
  let l = 0;
  let i = 0;

  while (l >= 0 && l < hops.length) {
    i++;
    l += hops[l]++;
  }

  return i;
}

// Part 2
function part2() {
  let hops = input.slice();
  let l = 0;
  let i = 0;

  while (l >= 0 && l < hops.length) {
    i++;
    let p = l;
    l += hops[l];
    hops[p] += (hops[p] >= 3 ? -1 : 1 );
  }

  return i;
}

module.exports = { part1, part2 }
